### DISCUSSION TITLE
How to keep contents intact when changing the route in ``single_page_example``?

### DISCUSSION THREAD

**Discussion by KiddoV** [zauberzeug/nicegui#2638] [Q&A]
### Question

This [SPA example](https://github.com/zauberzeug/nicegui/tree/main/examples/single_page_app) will keep re-initialize the entire page when changing the tab/button.

How do I change such behavior so my ``vue element`` will not call ``beforeUnmount()`` method when switching route?

Thanks!

**Comment by rodja** [comment]
Thanks for reaching out @KiddoV. But I do not quite understand the request yet. 

> How to keep contents intact when changing the route

Is not showing different content the purpose of changing the route? Maybe you could provide a simple and short example of what you have tried and where the approach is not matching your expectations.

**Reply by KiddoV** [reply]
Sorry  for the confusion.

Here is what I meant. For example if you have an input field on the first page:
```py
    @router.add('/')
    def show_one():
        ui.label('Content One').classes('text-2xl')
        ui.input("Type something...")
```
When you type something and leave the first page, then comeback, everything is gone:
<img alt="example" src="https://github.com/zauberzeug/nicegui/assets/28552977/e25ce8ce-08be-43ce-850a-75587c7c3372" width="420" />

I mean, a single page app would not behave like that.
How do I change this behavior and make it like a... tab routing?

**Reply by falkoschindler** [reply]
You could bind against user storage: https://nicegui.io/documentation/section_binding_properties#bind_to_storage

**Reply by KiddoV** [reply]
Hmmm, I mean this is just a small example. Imagine if a page have an entire ``editor element`` like MonacoEditor with thousand line of strings, or even a map with ``Leaflet element``, storing user interactive data manually would be pain.

**Reply by rodja** [reply]
If you switch routes, the old content is cleared and the function of the selected page executed. It will eat much more memory in the browser but you could save the old content in a "history" and turn just turn it invisible instead of clearing:

https://github.com/zauberzeug/nicegui/blob/38fcfe13fd8e5b297a8cb12aa357f94ad3d84eda/examples/single_page_app/router.py#L40

Then only build content if it is not already in the "history". Otherwise simply turn it visible again.

**Reply by mahesh11T** [reply]
same issue, I have a progress bar with task running in subprocess and when the page reloads or timeout, the background task continues but the progress bar resets to zero and elements too, how to handle this?

**Reply by falkoschindler** [reply]
> when the page reloads or timeout

This sounds like a different problem. This discussion is about switching routes in a single-page application. For problems with reloading pages or timeouts, I'd like to ask you to open a new Q&A with a clear problem statement and (ideally) a minimum reproducible code example. Thanks!

**Comment by rodja** [comment]
@Alyxion has started a promising pull request to implement generic SPA support in #2811. But still, the page content is cleared and recreated so inputs are lost if they are not stored somewhere. We are currently working on two more storage types which are be beneficial for SPAs: [`app.storage.tab`](https://github.com/zauberzeug/nicegui/pull/2837) which provides a place to store data across reloads in a single browser tab and [`app.storage.client`](https://github.com/zauberzeug/nicegui/pull/2820) which holds data only for the length of the current client connection (eg. will drop on every page change).
